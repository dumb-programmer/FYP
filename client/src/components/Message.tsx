import { TrashIcon } from "@heroicons/react/16/solid";
import CopyToClipboard from "./CopyToClipboardButton";

type MessageProps = {
    message: any;
    onDelete?: () => void;
}

export default function Message({ message, onDelete }: MessageProps) {
    return <div key={message._id} className="flex flex-col gap-8">
        <div className="flex items-center gap-4">
            <div className="h-20 w-20 bg-accent rounded-full flex items-center justify-center text-slate-50 font-bold">User</div>
            <p className="flex-1">{message.prompt}</p>
        </div>
        <div className="flex items-center gap-4">
            <div className="h-20 w-20 bg-info rounded-full flex items-center justify-center text-slate-50 font-bold">AI</div>
            <p className="flex-1">{message.response}</p>
        </div>
        {
            onDelete && <div className="flex justify-start pl-20 gap-2">
                <button className="btn btn-sm btn-ghost" onClick={onDelete}><TrashIcon color="red" height={15} width={15} /></button>
                <CopyToClipboard data={message.response} />
            </div>
        }
    </div>
}