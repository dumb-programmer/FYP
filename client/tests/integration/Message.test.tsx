import React from "react";
import { cleanup, render, screen, fireEvent } from "@testing-library/react";
import { expect, describe, it, vi, beforeAll, afterAll } from "vitest";
import Message from "../../src/components/Message";

const _ID = crypto.randomUUID();
const PROMPT = "Hello";
const RESPONSE = "Hi";

const onDeleteMock = vi.fn();
const onThumbsUpMock = vi.fn();
const onThumbsDownMock = vi.fn();
const renderComponent = (cursor = false, omitOnDelete = false, hasFeedback = false) => render(<Message message={{ _id: _ID, prompt: PROMPT, response: RESPONSE, feedback: hasFeedback ? { feedback: { positive: true } } : undefined }} onDelete={omitOnDelete ? undefined : onDeleteMock} onThumbsUp={onThumbsUpMock} onThumbsDown={onThumbsDownMock} cursor={cursor} />);

describe("Message", () => {

    beforeAll(() => {
        Object.assign(navigator, {
            clipboard: {
                writeText: vi.fn()
            }
        })
    });

    afterAll(() => {
        cleanup();
        vi.resetAllMocks();
    });

    it("renders correctly", () => {
        renderComponent();

        const component = screen.getByTestId("message");

        expect(component).toMatchSnapshot()
    });

    it("copy-to-clipboard button copies the response to clipboard", () => {
        renderComponent();

        const copyToClipboardButton = screen.getByTitle(/copy-to-clipboard/);

        fireEvent.click(copyToClipboardButton);

        expect(navigator.clipboard.writeText).toHaveBeenCalled();
        expect(navigator.clipboard.writeText).toHaveBeenCalledWith(RESPONSE);
    });

    it("onDelete is called when the delete button is clicked", () => {
        renderComponent();

        const deleteButton = screen.getByTitle(/delete-button/);

        fireEvent.click(deleteButton);

        expect(onDeleteMock).toHaveBeenCalled();
    });

    it("onThumbsUp is called on thumbs up button", () => {
        renderComponent();

        const thumbsUpBtn = screen.getByTitle(/thumbs-up/);

        fireEvent.click(thumbsUpBtn);

        expect(onThumbsUpMock).toHaveBeenCalled();
    });

    it("onThumbsDown is called on thumbs down button", () => {
        renderComponent();

        const thumbsDownBtn = screen.getByTitle(/thumbs-down/);

        fireEvent.click(thumbsDownBtn);

        expect(onThumbsDownMock).toHaveBeenCalled();
    });

    it("if the message already has feedback, then feedback buttons aren't clicked", () => {
        onThumbsUpMock.mockReset();
        onThumbsDownMock.mockReset();

        renderComponent(false, false, true);

        const thumbsUpBtn = screen.getByTitle(/thumbs-up/);

        fireEvent.click(thumbsUpBtn);

        expect(onThumbsUpMock).not.toHaveBeenCalled();

        const thumbsDownBtn = screen.getByTitle(/thumbs-down/);

        fireEvent.click(thumbsDownBtn);

        expect(onThumbsDownMock).not.toHaveBeenCalled();
    });

    it("cursor is not shown by default", () => {
        renderComponent();

        const cursorIcon = screen.queryByTestId(/cursor-icon/);

        expect(cursorIcon).not.toBeInTheDocument();
    });

    it("cursor is shown, if cursor props is true", () => {
        renderComponent(true);

        const cursorIcon = screen.getByTestId(/cursor-icon/);

        expect(cursorIcon).toBeInTheDocument();
    });

})