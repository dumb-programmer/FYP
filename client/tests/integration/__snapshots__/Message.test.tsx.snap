// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Message > renders correctly 1`] = `
<div
  class="flex flex-col gap-8"
  data-testid="message"
>
  <div
    class="flex items-center gap-4"
  >
    <div
      class="h-20 w-20 bg-accent rounded-full flex items-center justify-center text-slate-50 font-bold"
    >
      User
    </div>
    <p
      class="flex-1"
    >
      Hello
    </p>
  </div>
  <div
    class="flex items-center gap-4"
  >
    <div
      class="h-20 w-20 bg-info rounded-full flex items-center justify-center text-slate-50 font-bold"
    >
      AI
    </div>
    <p
      class="relative flex-1"
    >
      Hi
    </p>
  </div>
  <div
    class="flex justify-start pl-20 gap-2"
  >
    <button
      class="btn btn-sm btn-ghost"
      title="delete-button"
    >
      <svg
        aria-hidden="true"
        color="red"
        data-slot="icon"
        fill="currentColor"
        height="15"
        viewBox="0 0 16 16"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M5 3.25V4H2.75a.75.75 0 0 0 0 1.5h.3l.815 8.15A1.5 1.5 0 0 0 5.357 15h5.285a1.5 1.5 0 0 0 1.493-1.35l.815-8.15h.3a.75.75 0 0 0 0-1.5H11v-.75A2.25 2.25 0 0 0 8.75 1h-1.5A2.25 2.25 0 0 0 5 3.25Zm2.25-.75a.75.75 0 0 0-.75.75V4h3v-.75a.75.75 0 0 0-.75-.75h-1.5ZM6.05 6a.75.75 0 0 1 .787.713l.275 5.5a.75.75 0 0 1-1.498.075l-.275-5.5A.75.75 0 0 1 6.05 6Zm3.9 0a.75.75 0 0 1 .712.787l-.275 5.5a.75.75 0 0 1-1.498-.075l.275-5.5a.75.75 0 0 1 .786-.711Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <button
      class="btn btn-sm btn-ghost"
      title="copy-to-clipboard"
    >
      <svg
        aria-hidden="true"
        color="#696969"
        data-slot="icon"
        fill="currentColor"
        height="15"
        viewBox="0 0 16 16"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          clipboard-document-icon
        </title>
        <path
          clip-rule="evenodd"
          d="M11.986 3H12a2 2 0 0 1 2 2v6a2 2 0 0 1-1.5 1.937v-2.523a2.5 2.5 0 0 0-.732-1.768L8.354 5.232A2.5 2.5 0 0 0 6.586 4.5H4.063A2 2 0 0 1 6 3h.014A2.25 2.25 0 0 1 8.25 1h1.5a2.25 2.25 0 0 1 2.236 2ZM10.5 4v-.75a.75.75 0 0 0-.75-.75h-1.5a.75.75 0 0 0-.75.75V4h3Z"
          fill-rule="evenodd"
        />
        <path
          d="M3 6a1 1 0 0 0-1 1v7a1 1 0 0 0 1 1h7a1 1 0 0 0 1-1v-3.586a1 1 0 0 0-.293-.707L7.293 6.293A1 1 0 0 0 6.586 6H3Z"
        />
      </svg>
    </button>
    <button
      class="btn btn-sm btn-ghost disabled:bg-transparent"
      title="thumbs-up"
    >
      <svg
        aria-hidden="true"
        color="#696969"
        data-slot="icon"
        data-testid="thumbs-up-icon"
        fill="none"
        height="15"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.633 10.25c.806 0 1.533-.446 2.031-1.08a9.041 9.041 0 0 1 2.861-2.4c.723-.384 1.35-.956 1.653-1.715a4.498 4.498 0 0 0 .322-1.672V2.75a.75.75 0 0 1 .75-.75 2.25 2.25 0 0 1 2.25 2.25c0 1.152-.26 2.243-.723 3.218-.266.558.107 1.282.725 1.282m0 0h3.126c1.026 0 1.945.694 2.054 1.715.045.422.068.85.068 1.285a11.95 11.95 0 0 1-2.649 7.521c-.388.482-.987.729-1.605.729H13.48c-.483 0-.964-.078-1.423-.23l-3.114-1.04a4.501 4.501 0 0 0-1.423-.23H5.904m10.598-9.75H14.25M5.904 18.5c.083.205.173.405.27.602.197.4-.078.898-.523.898h-.908c-.889 0-1.713-.518-1.972-1.368a12 12 0 0 1-.521-3.507c0-1.553.295-3.036.831-4.398C3.387 9.953 4.167 9.5 5 9.5h1.053c.472 0 .745.556.5.96a8.958 8.958 0 0 0-1.302 4.665c0 1.194.232 2.333.654 3.375Z"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
    <button
      class="btn btn-sm btn-ghost disabled:bg-transparent"
      title="thumbs-down"
    >
      <svg
        aria-hidden="true"
        color="#696969"
        data-slot="icon"
        data-testid="thumbs-down-icon"
        fill="none"
        height="15"
        stroke="currentColor"
        stroke-width="1.5"
        viewBox="0 0 24 24"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.498 15.25H4.372c-1.026 0-1.945-.694-2.054-1.715a12.137 12.137 0 0 1-.068-1.285c0-2.848.992-5.464 2.649-7.521C5.287 4.247 5.886 4 6.504 4h4.016a4.5 4.5 0 0 1 1.423.23l3.114 1.04a4.5 4.5 0 0 0 1.423.23h1.294M7.498 15.25c.618 0 .991.724.725 1.282A7.471 7.471 0 0 0 7.5 19.75 2.25 2.25 0 0 0 9.75 22a.75.75 0 0 0 .75-.75v-.633c0-.573.11-1.14.322-1.672.304-.76.93-1.33 1.653-1.715a9.04 9.04 0 0 0 2.86-2.4c.498-.634 1.226-1.08 2.032-1.08h.384m-10.253 1.5H9.7m8.075-9.75c.01.05.027.1.05.148.593 1.2.925 2.55.925 3.977 0 1.487-.36 2.89-.999 4.125m.023-8.25c-.076-.365.183-.75.575-.75h.908c.889 0 1.713.518 1.972 1.368.339 1.11.521 2.287.521 3.507 0 1.553-.295 3.036-.831 4.398-.306.774-1.086 1.227-1.918 1.227h-1.053c-.472 0-.745-.556-.5-.96a8.95 8.95 0 0 0 .303-.54"
          stroke-linecap="round"
          stroke-linejoin="round"
        />
      </svg>
    </button>
  </div>
</div>
`;
